src/buildADTrecursion.C

```diff
@@ -0,0 +1,118 @@
+/* modify it under the terms of the GNU Lesser General Public */
+/* License as published by the Free Software Foundation; either */
+/* version 2.1 of the License, or (at your option) any later version. */
+
+/* This library is distributed in the hope that it will be useful, */
+/* but WITHOUT ANY WARRANTY; without even the implied warranty of */
+/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU */
+/* Lesser General Public License for more details. */
+
+/* You should have received a copy of the GNU Lesser General Public */
+/* License along with this library; if not, write to the Free Software */
+/* Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
+ */
+#include "buildADTrecursion.h"
+#include "codetypes.h"
+#include <algorithm>
+
+void buildADTrecursion(double *coord, double *adtReals, double *adtWork,
+                       int *adtIntegers, int *elementsAvailable, int *adtCount,
+                       int side, int parent, int level, int ndim, int nelem,
+                       int nav) {
+
+  int nd = ndim / 2;
+  int i, j;
+  int dimcut;
+  int nleft;
+  int ii, iip, jj, jjp;
+  int parentToChild;
+
+  if (nav > 1) {
+    //
+    // find the dimension to create the cut
+    //
+    dimcut = (level % ndim);
+    //
+    // collect coordinates along the dimension dimcut
+    //
+    for (i = 0; i < nav; i++)
+      adtWork[i] = coord[ndim * elementsAvailable[i] + dimcut];
+    //
+    // reorder elements with nleft elements to
+    // the left of median of adtWork
+    //
+    std::nth_element(elementsAvailable, elementsAvailable + (nav / 2),
+                     elementsAvailable + nav, [&](const int &a, const int &b) {
+                       return (adtWork[a] < adtWork[b]);
+                     });
+
+    nleft = (nav + 1) / 2;
+    (*adtCount)++;
+    ii = (*adtCount) * 4;
+    adtIntegers[ii] = elementsAvailable[nleft - 1];
+    adtIntegers[ii + 1] = -1;
+    adtIntegers[ii + 2] = -1;
+    adtIntegers[ii + 3] = -1;
+    //
+    // find minimum and maximum bounds of the elements
+    // contained in this leaf
+    //
+    for (i = 0; i < nd; i++) {
+      adtReals[ndim * (*adtCount) + i] = BIGVALUE;
+      adtReals[ndim * (*adtCount) + i + nd] = -BIGVALUE;
+    }
+    //
+    for (i = 0; i < nav; i++)
+      for (j = 0; j < nd; j++) {
+        ii = ndim * (*adtCount) + j;
+        iip = ii + nd;
+        jj = ndim * elementsAvailable[i] + j;
+        jjp = jj + nd;
+        //
+        adtReals[ii] = TIOGA_MIN(adtReals[ii], coord[jj]);
+        adtReals[iip] = TIOGA_MAX(adtReals[iip], coord[jjp]);
+      }
+    //
+    // specify that the new element is the child of parent
+    // unless root
+    //
+    if (side > 0) {
+      adtIntegers[4 * parent + side] = elementsAvailable[nleft - 1];
+    }
+    parentToChild = *adtCount;
+    //
+    // build the left side of the tree
+    //
+    if (nleft > 1) {
+      buildADTrecursion(coord, adtReals, adtWork, adtIntegers,
```

Comment:
@jrood-nrel adtWork is used here and below too.

Reply:
It's overwritten entirely each call to `buildADTrecursion` so sorting it looks unnecessary.